TURN relay implementation

longfin opened the pull request:
This PR introduces [TURN](https://tools.ietf.org/html/rfc5766) client as relay mechanism to accomplish [NAT traversal](https://en.wikipedia.org/wiki/NAT_traversal). (see also #73) it consists of three parts as below.

- Added some helpers
  - CRC32
  - bytes ↔ uint (for masking) converter
  - Proxy for NetworkStream
- Added STUN / TURN message protocols
  - STUN / TURN attributes
  - STUN / TURN messages
- Added TURN protocol client

Currently, it doesn't have any open interface. so I'll add `[skip changelog]`. (perhaps it is necessary to write in the next PR.)

# TODO
- [x] Make the test case optional
- [x] ~~Fill STUN / TURN enum description.~~, I'll handle this at next PR.

dahlia commented on Libplanet/Net/Stun/TurnClient.cs:
```diff
@@ -0,0 +1,154 @@
+            }
+            else
+            {
+                throw new TurnClientException("Allocate Failed.", response);
```
Although it's trivial, F in the second word should be small, I think?

dahlia commented on Libplanet/Net/Stun/TurnClient.cs:
```diff
@@ -0,0 +1,154 @@
+{
+    internal class TurnClient : IStunContext, IDisposable
+    {
+        private const int AllocateRetry = 5;
```
Is this an arbitrary number we decided?

longfin replied:
Yes, it's an arbitrary.

dahlia commented on Libplanet/Net/Stun/Messages/ConnectRequest.cs:
```diff
@@ -0,0 +1,21 @@
+{
+    internal class ConnectRequest : StunMessage
+    {
+        public ConnectRequest(IPEndPoint peerAddr)
```
We'd better to have a consistent term for `IPEndPoint`.  Sometimes we name `IPEndPoint` values “address” and sometimes “end point.”

dahlia commented on Libplanet/Net/Stun/Messages/CreatePermissionSucessResponse.cs:
```diff
@@ -0,0 +1,9 @@
+namespace Libplanet.Net.Stun.Messages
+{
+    internal class CreatePermissionSucessResponse : StunMessage
```
A typo here.  S should be twice in “sucess” → “success.”

ipdae commented:
😢 Window build is broken.

longfin commented:
> * How long would it take if we separate `Libplanet.Net.Strun`'s own project from Libplanet project?

Can't make sure, but I'll try on next PR.

> * Could we omit those `Can` prefixes from test methods when it's appropriate?

Actually, I'm confused about the naming convention for test cases 😕 .

longfin commented:
I've rebased to apply reviews.

And, I removed `Can` prefix in test cases. (I'm not sure about the convention yet, but I'm sure it's more natural than having a `Can`)
